File: src/main/java/cz/linkedlist/AbstractTileDownloader.java
package cz.linkedlist;

import org.springframework.beans.factory.annotation.Value;

import java.io.File;
import java.util.List;

/**
 * @author Martin Macko <https://github.com/LinkedList>
 */
public abstract class AbstractTileDownloader implements TileDownloader {
    @Value(DESTINATION_FOLDER_PROP)
    protected String destinationFolder;


    @Override
    public void downBand(TileSet tileSet, int band) {
        tileSet.band(band)
                .map(s -> down(tileSet, s))
                .orElseThrow(() -> new RuntimeException("I am sorry, cannot download band: " + band));
    }

    @Override
    public void downBand8A(TileSet tileSet) {
        down(tileSet, tileSet.band8A());
    }

    @Override
    public void downProductInfo(TileSet tileSet) {
        down(tileSet, tileSet.productInfo());
    }

    @Override
    public void downTileInfo(TileSet tileSet) {
        down(tileSet, tileSet.tileInfo());
    }

    @Override
    public void downMetadata(TileSet tileSet) {
        down(tileSet, tileSet.metadata());
    }

    @Override
    public void downContent(TileSet tileSet, TileSet.Contents content) {
        this.down(tileSet, content.map(tileSet));
    }

    @Override
    public void downContent(TileSet tileSet, List<TileSet.Contents> contents) {
        contents.forEach(c -> this.downContent(tileSet, c));
    }

    protected abstract File down(TileSet tileSet, String what);

    public void setDestinationFolder(String destinationFolder) {
        this.destinationFolder = destinationFolder;
    }
}

File: src/main/kotlin/cz/linkedlist/AbstractTileDownloader.kt
package cz.linkedlist

import org.springframework.beans.factory.annotation.Value

import java.io.File

/**
 * @author Martin Macko <https:></https:>//github.com/LinkedList>
 */
abstract class AbstractTileDownloader : TileDownloader {
    @Value(TileDownloader.DESTINATION_FOLDER_PROP)
    lateinit var destinationFolder: String


    override fun downBand(tileSet: TileSet, band: Int) {
        tileSet.band(band)
            .map { s -> down(tileSet, s) }
            .orElseThrow { RuntimeException("I am sorry, cannot download band: $band") }
    }

    override fun downBand8A(tileSet: TileSet) {
        down(tileSet, tileSet.band8A())
    }

    override fun downProductInfo(tileSet: TileSet) {
        down(tileSet, tileSet.productInfo())
    }

    override fun downTileInfo(tileSet: TileSet) {
        down(tileSet, tileSet.tileInfo())
    }

    override fun downMetadata(tileSet: TileSet) {
        down(tileSet, tileSet.metadata())
    }

    override fun downContent(tileSet: TileSet, content: TileSet.Contents) {
        this.down(tileSet, content.map(tileSet))
    }

    override fun downContent(tileSet: TileSet, contents: Array<TileSet.Contents>) {
        contents.forEach { c -> this.downContent(tileSet, c) }
    }

    protected abstract fun down(tileSet: TileSet, what: String): File


    override fun isDownloaded(what: String): File? {
        val file = File(destinationFolder + what.replace("/", "_"))
        return if (file.exists()) file else null
    }
}
